server/plugin_release.go

```diff
@@ -230,170 +443,275 @@ func downloadAsset(assetURL, tempFolder string) (string, error) {
 	// Write the body to file
 	_, err = io.Copy(out, resp.Body)
 	if err != nil {
-		return "", err
+		return "", nil
 	}
-	LogInfo("Done downloading")
-	return filepath, nil
+
+	return pathToFile, nil
 }
 
-func copyFileToSigningServer(fileToCopy, tempFolder string) (string, error) {
-	LogInfo("Will copy the artifact to the signing server")
-	clientConfig, _ := auth.PrivateKey(Cfg.PluginSigningSSHUser, Cfg.PluginSigningSSHKeyPath, ssh.InsecureIgnoreHostKey())
-	host := fmt.Sprintf("%s:22", Cfg.PluginSigningSSHHost)
+// getPluginAsset polls till it finds the plugin tar file.
+func getPluginAsset(ctx context.Context, githubClient *github.Client, owner, repo, tag string) (*github.ReleaseAsset, error) {
+	LogInfo("Checking if the release asset is available")
 
-	clientConfig.Timeout = 30 * time.Minute
-	client := scp.NewClientWithTimeout(host, &clientConfig, 30*time.Minute)
+	ctx, cancel := context.WithTimeout(ctx, 10*time.Minute)
+	defer cancel()
 
-	err := client.Connect()
-	if err != nil {
-		LogError("Couldn't establish a connection to the remote server. err=" + err.Error())
-		return "", err
-	}
+	for {
+		// Using timer to avoid memory leaks
+		timer := time.NewTimer(30 * time.Second)
+		defer timer.Stop()
 
-	f, _ := os.Open(fileToCopy)
-	defer client.Close()
-	defer f.Close()
+		release, _, err := githubClient.Repositories.GetReleaseByTag(ctx, owner, repo, tag)
```

Comment:
@jupenur
you mean the strings that are being passed here?

Can't we treat this as a proxy and assume that the server will reject any `../` sequences on their end? Unless i'm missing something.

Reply:
> you mean the strings that are being passed here?

Right.

>Can't we treat this as a proxy and assume that the server will reject any `../` sequences

Server, as in `api.github.com`? No we can't.

This is what `go-github` is doing internally: First it builds a path string using `fmt.Sprintf`...

```go
u := fmt.Sprintf("repos/%s/%s/releases/tags/%s", owner, repo, tag)
```

Then it passes that to `url.Parse`:

```go
	u, err := c.BaseURL.Parse(urlStr)
```

So if we pass a `repo` value of `../jupenur/evilrepo` to `go-github`, it will first turn it into a string like `repos/mattermost/../jupenur/evilrepo/releases/tags/v1.0.0`, then parse it as a URL. And because `url.Parse` automatically collapses `../` sequences, the end result is `https://api.github.com/repos/jupenur/evilrepo/releases/tags/v1.0.0`. This is the URL that gets requested from the server, so GitHub never sees a single `../` sequence, it just returns a repo from the wrong owner.

Even if the `../` sequence did get sent to GitHub, it would likely collapse it itself. And if it didn't, we still couldn't rely on that behavior as a security control, since GitHub could decide to change it at any time.

Reply:
Maybe I don't understand the issue, but would it make sense to solve it in the go-github lib?

Reply:
Possibly.